apps/web-mzima-client/src/app/core/guards/deactivate-settings-route.guard.ts

```diff
@@ -0,0 +1,34 @@
+import { Injectable } from '@angular/core';
+import { CanDeactivate } from '@angular/router';
+import { Observable } from 'rxjs';
+
+export interface IDeactivateGuard {
+  changesMade: boolean | null;
+  modalConfirmButtonClicked: boolean;
+  browserBackButtonClicked: boolean;
+  submitted: boolean;
+  openConfirmModal: ({ navigateBack }: { navigateBack: boolean }) => void;
+}
+
+@Injectable({ providedIn: 'root' })
+export class DeactivateGuardService implements CanDeactivate<IDeactivateGuard> {
+  canDeactivate(component: IDeactivateGuard): boolean | Promise<boolean> | Observable<boolean> {
+    const formHasChangedButModalConfirmButtonNotClicked =
+      component?.changesMade && !component.modalConfirmButtonClicked;
+
+    const guardDetects = {
+      cancelButtonOrRouteLinkClick:
+        formHasChangedButModalConfirmButtonNotClicked &&
+        !component.browserBackButtonClicked &&
+        !component.submitted,
+      browserBackButtonClick:
+        formHasChangedButModalConfirmButtonNotClicked && component.browserBackButtonClicked,
+    };
+
+    if (guardDetects.cancelButtonOrRouteLinkClick || guardDetects.browserBackButtonClick) {
+      component.openConfirmModal({ navigateBack: component.modalConfirmButtonClicked });
```

Comment:
You don't need all the tracking of buttons clicked in the guard, just check for changesMade, if its true, open the modal from here (we need to make the language more general so we can use it in many components, but I think that is fine.) You need to add something like this to the canDeactivateGuard: 
```  if (component?.changesMade) {
      return this.openConfirmModal().then((confirmed: any) => confirmed);
    } else {
      return true;
    }
```

And move the openConfirmModal()-function to the actual guard.

Reply:
It wasn't as straight forward as just checking for `changesMade` only.
The part about moving the confirm modal to the guard is clear. When I get to that point, I will be able to explain better on slack. For now, I will do the best I can.